optimizely/event/dispatcher.go

```diff
@@ -54,10 +55,72 @@ func (*HTTPEventDispatcher) DispatchEvent(event LogEvent, callback func(bool)) {
 		if resp.StatusCode == 204 {
 			success = true
 		} else {
-			fmt.Printf("http.Post invalid response %d", resp.StatusCode)
+			dispatcherLogger.Error(fmt.Sprintf("http.Post invalid response %d", resp.StatusCode), nil)
 			success = false
 		}
 	}
 	callback(success)
+}
+
+// QueueEventDispatcher is a queued version of the event dispatcher that queues, returns success, and dispatches events in the background
+type QueueEventDispatcher struct {
+	eventQueue     Queue
+	eventFlushLock sync.Mutex
+	dispatcher     *HTTPEventDispatcher
+}
+
+// DispatchEvent queues event with callback and calls flush in a go routine.
+func (ed *QueueEventDispatcher) DispatchEvent(event LogEvent, callback func(bool)) {
+
+	ed.eventQueue.Add(event)
+	callback(true)
+	go func() {
+		ed.flushEvents()
+	}()
+}
+
+// flush the events
+func (ed *QueueEventDispatcher) flushEvents() {
+
+	ed.eventFlushLock.Lock()
+
+	defer func(){
+		ed.eventFlushLock.Unlock()
+	}()
+
+	for ed.eventQueue.Size() > 0 {
+		items := ed.eventQueue.Get(1)
+		if len(items) == 0 {
+			time.Sleep(5 * time.Second)
+			continue
+		}
+		event, ok := items[0].(LogEvent)
+		if !ok {
+			// remove it
+			dispatcherLogger.Error("invalid type passed to event dispatcher", nil)
+			ed.eventQueue.Remove(1)
+			continue
+		}
+
+		ed.dispatcher.DispatchEvent(event, func(success bool) {
+			if success {
+				ed.eventQueue.Remove(1)
+			} else {
+				// we failed.  Sleep 5 seconds and try again.  Or, should we exit and try the next time a log event is added?
+				time.Sleep(5 * time.Second)
+			}
+		})
+	}
+}
+
+// NewQueueEventDispatcher creates a dispatcher that queues in memory and then sends via go routine.
+func NewQueueEventDispatcher(ctx context.Context) Dispatcher {
+	dispatcher := &QueueEventDispatcher{eventQueue: NewInMemoryQueue(1000), dispatcher:&HTTPEventDispatcher{}}
+
+	go func() {
+		<-ctx.Done()
```

Comment:
Is this needed if you call flush events from the parent?

Reply:
I'm not sure it is needed.  But, if there are no events to send it is pretty much a noop.

Reply:
is canceled context going to stop event dispatcher too ?

Reply:
please look at https://github.com/optimizely/go-sdk/blob/master/optimizely/config/polling_manager.go#L54 - context will stop the whole mechanism

Reply:
I think this will need to be revised to account for the wait group implementation that @pawels-optimizely is working

Reply:
It's fine to merge this in and have Pawel add the wait group here afterwards